### PR TITLE
Exposing OkHttpClient.Builder for the app to plug in their own interceptors

### PR DESCRIPTION
**Pull request by bhariharan**
This adds the ability to plug a `TwinPrime` client in.

### CODE REVIEW

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java [hunk 1]
```diff
@@ -148,35 +157,52 @@ public static void clearOkClientsCache() {
...
 	/**
-	 * Sets the OkHttpclient associated with this user account. The OkHttpclient
+	 * Sets the OkHttpclient.Builder associated with this user account. The OkHttpclient.Builder
 	 * are cached in a map and reused as and when a user account
 	 * switch occurs, to prevent multiple threads being spawned unnecessarily.
 	 */
-	private synchronized void setOkHttpClient() {
-		if (OK_CLIENTS == null) {
-			OK_CLIENTS = new HashMap<>();
+	private synchronized void setOkHttpClientBuilder() {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java]: Building a map of builders.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java [hunk 2]
```diff
@@ -148,35 +157,52 @@ public static void clearOkClientsCache() {
...
-		this.okHttpClient = okHttpClient;
+	public synchronized void setOkHttpClient(OkHttpClient okHttpClient) {
+		if (OK_CLIENTS == null) {
+			OK_CLIENTS = new HashMap<>();
+		}
+		final String uniqueId = this.oAuthRefreshInterceptor.clientInfo.buildUniqueId();
+		OkHttpClient curOkHttpClient = null;
+		if (uniqueId != null) {
+			curOkHttpClient = OK_CLIENTS.get(uniqueId);
+			if (curOkHttpClient == null) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java]: If there's no existing client in the map, check if the passed in one is valid. If not, build a new one. If the passed in one is valid, replace the existing one with the new one. If there is a valid one and the passed in one is invalid, do nothing.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java [hunk 3]
```diff
@@ -148,35 +157,52 @@ public static void clearOkClientsCache() {
...
+				if (okHttpClient == null) {
+					curOkHttpClient = getOkHttpClientBuilder().build();
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java]: Getting the builder object from the map built above.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java [hunk 4]
```diff
@@ -242,6 +268,13 @@ public ClientInfo getClientInfo() {
 	}
 
 	/**
+	 * @return underlying OkHttpClient.Builder
+	 */
+	public OkHttpClient.Builder getOkHttpClientBuilder() {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java]: Exposing builder object for apps.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java [hunk 5]
```diff
@@ -148,35 +157,52 @@ public static void clearOkClientsCache() {
...
+	 * switch occurs, to prevent multiple threads being spawned unnecessarily.
 	 */
-	public void setOkHttpClient(OkHttpClient okHttpClient) {
-		this.okHttpClient = okHttpClient;
+	public synchronized void setOkHttpClient(OkHttpClient okHttpClient) {
+		if (OK_CLIENTS == null) {
+			OK_CLIENTS = new HashMap<>();
+		}
+		final String uniqueId = this.oAuthRefreshInterceptor.clientInfo.buildUniqueId();
+		OkHttpClient curOkHttpClient = null;
```
Comment (wmathurin) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java]: That method is a bit hard to follow, could we simplify it e.g.

``` java
    public synchronized void setOkHttpClient(OkHttpClient okHttpClient) {
    
        if (OK_CLIENTS == null) {
            OK_CLIENTS = new HashMap<>();
        }

        final String uniqueId = this.oAuthRefreshInterceptor.clientInfo.buildUniqueId();

        if (uniqueId == null) {
            return; // what now?
        }

        // Valid client passed in, cache it
        if (okHttpClient != null) {
            OK_CLIENTS.put(uniqueId, okHttpClient);
        }
        // Null client passed in, create and cache one if we don't have one cached
        else if (!OK_CLIENTS.containsKey(uniqueId)) {
            OK_CLIENTS.put(uniqueId, getOkHttpClientBuilder().build());
        }
        
        // Use cached client
        this.okHttpClient = OK_CLIENTS.get(uniqueId);
    }
```

**Comment by bhariharan** [comment]
@wmathurin Done [here](https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/1371/commits/7ee3d602d0eace828b88003b85320b711cbb4621).

**Comment by bhariharan** [comment]
There's a single test failure which is unrelated to this pull request:
```
com.salesforce.androidsdk.phonegap.SmartSyncJSTest > testStoreCacheWithGlobalStoreNamed[hudson_en-US_160_WVGA_android-23_google_apis-x86(AVD) - 6.0] FAILED
        junit.framework.AssertionFailedError: testStoreCacheWithGlobalStoreNamed Timeout (30 seconds) exceeded
        at com.salesforce.androidsdk.phonegap.util.test.JSTestCase.runTest(JSTestCase.java:149)
```
Merging.
